clash-ghc/src-ghc/Clash/GHC/LoadInterfaceFiles.hs

```diff
@@ -168,145 +256,132 @@ loadIface foundMod = do
                                            ]
                          in traceIf True msg' (return Nothing)
 
-loadExternalBinders
-  :: GHC.GhcMonad m
-  => HDL
-  -> [CoreSyn.CoreBndr]
-  -> m LoadedBinders
+-- | Given a list of top-level binders, recursively load all the binders,
+-- primitives, and type classes it is using. (Exported function.)
+loadExternalBinders :: GHC.GhcMonad m => HDL -> [CoreSyn.CoreBndr] -> m LoadedBinders
 loadExternalBinders hdl bndrs = do
-  loaded <- mergeLoadedBinders <$> mapM (loadExprFromIface hdl) bndrs
-  fst <$>
-    loadExternalExprs'
-      hdl
-      loaded
-      (UniqSet.mkUniqSet (collectLbBinders loaded))
-      (map snd (lbBinders loaded))
-
-loadExternalExprs
-  :: GHC.GhcMonad m
-  => HDL
-  -> UniqSet.UniqSet CoreSyn.CoreBndr
-  -> [CoreSyn.CoreBind]
-  -> m LoadedBinders
-loadExternalExprs hdl = go emptyLb
-  where
-    go loaded _ [] =
-      return loaded
-    go loaded0 visited0 (CoreSyn.NonRec _ e:bs) = do
-      (loaded1, visited1) <- loadExternalExprs' hdl loaded0 visited0 [e]
-      go loaded1 visited1 bs
-    go loaded0 visited0 (CoreSyn.Rec bs:bs') = do
-      (loaded1, visited1) <- loadExternalExprs' hdl loaded0 visited0 (map snd bs)
-      go loaded1 visited1 bs'
-
--- | Used by entry points: 'loadExternalExprs', 'loadExternalBinders'
-loadExternalExprs'
-  :: GHC.GhcMonad m
-  => HDL
-  -> LoadedBinders
-  -> UniqSet.UniqSet CoreSyn.CoreBndr
-  -> [CoreSyn.CoreExpr]
-  -> m ( LoadedBinders, UniqSet.UniqSet CoreSyn.CoreBndr)
-loadExternalExprs' _hdl loaded visited [] =
-  return (loaded, visited)
-loadExternalExprs' hdl loaded0 visited0 (e:es) = do
-  let
-    isInteresting v =
-         Var.isId v
-      && not (v `UniqSet.elementOfUniqSet` visited0)
-      && isNothing (Id.isDataConId_maybe v)
-
-    fvs0 = CoreFVs.exprSomeFreeVarsList isInteresting e
-    fvs1 = map (\v -> maybe (Left v) (Right . (v,)) (Id.isClassOpId_maybe v)) fvs0
-    (fvs2, clsOps0) = partitionEithers fvs1
-    clsOps1 = map goClsOp clsOps0
-
-  loaded1 <- mergeLoadedBinders <$> mapM (loadExprFromIface hdl) fvs2
-
-  loadExternalExprs'
-    hdl
-    (mergeLoadedBinders [loaded0, loaded1, emptyLb{lbClassOps=clsOps1}])
-    (foldl' UniqSet.addListToUniqSet visited0 [collectLbBinders loaded1, map fst clsOps0])
-    (es ++ map snd (lbBinders loaded1))
+  flip execStateT emptyLb $ do
+    mapM_ (loadExprFromIface hdl) bndrs
+
+-- Given a list of binds, recursively load all its binders, primitives, and
+-- type classes it is using. (Exported function.)
+loadExternalExprs :: GHC.GhcMonad m => HDL -> [CoreSyn.CoreBind] -> m LoadedBinders
+loadExternalExprs hdl binds0 = do
+  flip execStateT initLb $ do
```

Comment:
Not one, but _two_ redundant `do`. I suppose this is what I get for having HLint in my IDE setup

Reply:
TBH, I like the `do`. It makes it easier to extend later on. But I'll remove it if it bothers you/hlint 👍